File: algorithms/03_cpp/sorting/counting_sort.cpp
/**
 * Counting sort for non-negative integers.
 */

#include <vector>
#include <algorithm>
#include <cassert>
#include <iostream>

void countingSort(std::vector<int>& a, int maxVal) {
    std::vector<int> count(maxVal + 1, 0);
    for (int x : a) count[x]++;
    int idx = 0;
    for (int v = 0; v <= maxVal; ++v)
        for (int c = 0; c < count[v]; ++c) a[idx++] = v;
}

int main() {
    std::vector<int> data = {4, 2, 2, 8, 3, 3, 1};
    countingSort(data, 8);
    assert(std::is_sorted(data.begin(), data.end()));
    assert(data.front() == 1 && data.back() == 8);
    std::cout << "[C++ CountingSort] Verified." << std::endl;
    return 0;
}

File: algorithms/04_java/sorting/CountingSort.java
import java.util.Arrays;

public class CountingSort {
    // Sorts non-negative integers in [0, maxVal].
    public static int[] sort(int[] a, int maxVal) {
        int[] count = new int[maxVal + 1];
        for (int x : a) count[x]++;
        for (int i = 1; i <= maxVal; i++) count[i] += count[i - 1];
        int[] out = new int[a.length];
        for (int i = a.length - 1; i >= 0; i--) out[--count[a[i]]] = a[i];
        return out;
    }

    public static void main(String[] args) {
        int[] data = {3, 1, 3, 0, 2, 1, 5, 4, 2};
        int[] sorted = sort(data, 5);
        for (int i = 1; i < sorted.length; i++)
            if (sorted[i - 1] > sorted[i]) throw new AssertionError("not sorted at " + i);
        System.out.println("[Java CountingSort] Counting sort verified: " + Arrays.toString(sorted));
    }
}
